keps/sig-storage/20191012-storagepool-advplacement.md

```diff
@@ -0,0 +1,364 @@
+
+#### Story 2
+
+Let’s assume a scale out storage service, e.g. a Minio Object Store cluster using Erasure Coding. The scope of the Minio cluster should be the same as the Kubernetes cluster, with 1 Minio Pod per Kubernetes Node. Minio is configured with erasure coding such that half the chunks are parity and the other half are data. Minio operates the erasure code on a per-disk basis and optimizes not just for host but also individual disk failures this way. Hence, let’s assume every Node has 4 physical drives. The SDS exposes 4 “placement choices” mapping the 4 physical disks to CSI topology choices. Minio then uses Pod affinity and Storage Class allowsTopology with the custom key com.example/disk=Disk1 for its first PVC, com.example/disk=Disk2 for its second PVC, etc., for a total of 4 PVCs matching the physical topology.
+
+The placement choices in this use case cannot be expressed using existing Kubernetes mechanism.
+
+![Story2-LocalDrives](./kep-storagepool-story2.png)
+
+Beyond the existence of the Placement Choice / Storage Pool, Minio further uses metadata to know that the storage pool has enough capacity to accommodate the desired PVC. If for example physical disks have different size, then Minio may decide to create PVCs of different sizes to still work in this scenario.
+
+#### Story 3
+
+Let’s assume a sharded and replicated MongoDB instance. Let’s further assume that the Cloud Storage Service used for the PVCs is comprised of 3 distinct Storage-Pods (not to be confused with K8s Pods) in both of the two available AZs, such that the 3 Storage-Pods fail independently by having dedicated power, networking, racks, hosts and disks. The CSI driver of the Cloud Storage service announces that 3 Storage-Pods per AZ as placement choices (AZ1-SP1, AZ1-SP2, AZ1-SP3, AZ2-SP1, etc.) and also advertises that all nodes in AZ1 can access all Storage-Pods in AZ1 (as a simple example). Let’s further assume MongoDB is configured for 3-way replication per shard. MongoDB decides to create 3 Pods for a given shard:
+* Pod1 has an affinity to AZ1, and its PVCs have an affinity to AZ1-SP1.
+* Pod2 has an affinity to AZ1, and its PVCs have an affinity to AZ1-SP2.
+* Pod3 has an affinity to AZ2, and its PVCs have an affinity to AZ2-SP1.
+
+If we compare this to the case where we can’t express the PVC affinity to the Storage-Pod, Pod1 and Pod2 may land on the same Storage-Pod and so if that Storage-Pod fails, it would take down two Pods. In the scenario described here, the math will show a higher availability and durability.
+
+The placement choices in this use case cannot be expressed using existing Kubernetes mechanism.
+
+![Story3-MultiStoragePods](./kep-storagepool-story3.png)
+
+
+### Implementation Details/Notes/Constraints
+
+#### New CRD to surface Storage Pools
+```
+apiVersion: storagepool.storage.k8s.io/v1alpha1
+kind: StoragePool
+metadata:
+  name: storagePool1
```

Comment:
Thats a good question. Do we have precedent for the infrastructure surfacing up entities and having to give unique names? I would assume PV names fall into that?

Reply:
PVs get unreadable names like `pvc-24fcddf6-9178-4fe4-b71f-4417652bca5d`. One has to find the right object through a reference elsewhere (`PVC.Spec.VolumeName` is set to it) or by searching for the right one by its attributes.

Reply:
OK, yeah, thats likely not a great path to follow :)

Reply:
@cdickmann - for load balancer services, kubernetes relies on service `uid` field to name/create resources at infrastructure provider level, also +1 on PVs

Reply:
In the "New external controller for StoragePool" section, we mentioned that CSI driver will implement a ControllerListStoragePools() function.  The controller will call that CSI function at startup to discover all pools which include names of the pools on the storage systems.  We should probably consider appending a uuid to those names to avoid name collision.